qiskit/assembler/assemble_circuits.py

```diff
@@ -105,18 +112,76 @@ def _assemble_circuit(circuit):
             del instruction._condition
 
         instructions.append(instruction)
-    return QasmQobjExperiment(instructions=instructions, header=header,
-                              config=config)
+    return (QasmQobjExperiment(instructions=instructions, header=header, config=config),
+            pulse_library)
+
+
+def _assemble_pulse_gates(circuit, run_config):
+    """Assemble and return the circuit calibrations and associated pulse library, if there are any.
+    The calibrations themselves may reference the pulse library which is returned as a dict.
+
+    Args:
+        circuit (QuantumCircuit): circuit which may have pulse calibrations
+        run_config (RunConfig): configuration of the runtime environment
+
+    Returns:
+        tuple(QasmExperimentCalibrations, dict(str, array)): The calibrations and pulse library.
+    """
+    if not circuit.calibrations:
+        return None, None
+    if not hasattr(run_config, 'parametric_pulses'):
+        run_config.parametric_pulses = []
+    calibrations = []
+    pulse_library = {}
+    for gate, cals in circuit.calibrations.items():
+        for (qubits, params), schedule in cals.items():
+            qobj_instructions, _ = assemble_schedule(
+                schedule,
+                converters.InstructionToQobjConverter(PulseQobjInstruction),
+                run_config,
+                pulse_library)
+            calibrations.append(
+                GateCalibration(str(gate), list(qubits), list(params), qobj_instructions))
+    return QasmExperimentCalibrations(gates=calibrations), pulse_library
+
+
+def _extract_common_calibrations(experiments):
+    """Given a list of ``QasmQobjExperiment``s, each of which may have calibrations in their
+    ``config``, collect common calibrations into a global ``QasmExperimentCalibrations``
+    and delete them from their local experiments.
+    """
+    if not (experiments and all(hasattr(exp.config, 'calibrations') for exp in experiments)):
+        return experiments, None
+
+    common_calibrations = []
+    from collections import defaultdict
+    gate_hash_map = defaultdict(list)
+    for exp_idx, exp in enumerate(experiments):
+        for gate_idx, gate_cal in enumerate(exp.config.calibrations.gates):
+            gate_hash_map[str(gate_cal.to_dict())].append((exp_idx, gate_idx, gate_cal))
+
+    # Collect common cals, remove from respective experiments
+    for gate_cal, exps_w_cal in gate_hash_map.items():
+        if len(exps_w_cal) == len(experiments):
+            common_calibrations.append(exps_w_cal[0][2])
```

Comment:
Seems like `exps_w_cal[0]` is just an integer, so there is no element `exps_w_cal[0][2]`.

Reply:
I agree it's confusing, `exps_w_cal[0]` is the first element in the list `exps_w_cal`, and each element is a tuple. The way it was originally written was rather terrible. The new version should be easier to read.


(In case it still seems odd, there are some subtle points:
1. It seems like it'd be nice to map `{GateCalibration(...): [exp0, exp1, exp2]}`, but identical gates map to different keys unless the hash of the gate is used, so it ends up `{hash_int_of_gate: [exp0, exp1, exp2]}`. Not what I originally expected.
2. When I later want to remove a GateCalibration from one experiment's list of calibrated gates, `.remove` requires that the object passed to `remove` `is` the object in the list. I also need a reference to one of the gates to collect for the top level cal, so I chose `{hash_int: [(exp0, gatecal), ...]}`.

The alternative would be to keep a mapping from `hash_int` back to one of the `gatecals`, and to loop through the experiment's calibrated gates to find and remove the matching one.

)